### PR TITLE
adding MlflowClient to top level mlflow package

### LINKED ISSUE #6080: [FR] [Roadmap] Alias client at top-level
## MLflow Roadmap Item

This is an MLflow Roadmap item that has been prioritized by the MLflow maintainers. We’ve identified this feature as a highly requested addition to the MLflow package based on community feedback. 
We're seeking a community contribution for the implementation of this feature and will enthusiastically support the development and review of a submitted PR for this.

## Contribution Note

As with other roadmap items, there may be a desire for multiple contributors to work on an issue. While we don’t discourage collaboration, we strongly encourage that a primary contributor is assigned to roadmap issues to simplify the merging process. 
The items on the roadmap are of a high priority. Due to the wide-spread demand of roadmap features, we encourage potential contributors to only agree to take on the work of creating a PR, making changes, and ensuring that test coverage is adequately created for the feature if they are willing and able to see the implementation through to a merged state. 

## Feature scope

This roadmap feature’s complexity is classified as:

- [X] `good-first-issue`: This feature is limited in complexity and effort required to implement.
- [ ] `simple`: This feature does not require a large amount of effort to implement and / or is clear enough to not need a design discussion with maintainers.
- [ ] `involved`: This feature will require a substantial amount of development effort but does not require an agreed-upon design from the maintainers. The feedback given during the PR phase may be involved and necessitate multiple iterations before approval. (Please bear with us as we collaborate with you to make a great contribution)
- [ ] `design-recommended`: This is a substantial feature that should have a design document approved prior to working on an implementation (to save your time, not ours). After agreeing to work on this feature, a maintainer will be assigned to support you throughout the development process.

## Proposal Summary

Create an alias at the top package level for `mlflow.tracking.MlflowClient()` to more accurately reflect its common usage and to simplify import and caller statements. 
The client constructor should be referable from: `mlflow.MlflowClient()`.

Note: this is simply adding an alias, not changing the package classpath to move the client constructor.

  Simplify the user experience.
> #### Why is this use case valuable to support for MLflow users in general? 
  Everyone (*citation needed) appreciates typing less.
> #### Why is it currently difficult to achieve this use case? 
  N/A

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by subramaniam02**
Signed-off-by: subramaniam02 <subramaniam02@gmail.com>

## Related Issues/PRs

Close #6080 

## What changes are proposed in this pull request?
- added MlflowClient at top level mlflow package
- changed all references of `from mlflow.tracking import MlflowClient`

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/__init__.py [hunk 1]
```diff
@@ -43,7 +43,6 @@
 import mlflow.projects as projects
 import mlflow.tracking as tracking
 import mlflow.models
-import mlflow.artifacts
```
Comment (BenWilson2) [inline mlflow/__init__.py:46]: was this a typo?

Comment (subramaniam02) [reply mlflow/__init__.py:46]: Not a typo. I thought it was not unused import but I was wrong. I am reverting this change. I am facing some import issues and will update the PR with more changes.

**Comment by subramaniam02** [comment]
@BenWilson2 PR is ready for review. Tests are running fine. I would like to get your thoughts on following
- There still a patch of code that is not part of the PR. Let me know if you want this patch to be part of the PR https://github.com/subramaniam02/mlflow/pull/3/commits/aa9bf4a74ee0aef4c7a34c365372b4e20e31f803
- Code has many versions of `MlflowClient` like `mlflow.MlflowClient()` and `MlflowClient()`. Do we want uniformity ?

**Comment by BenWilson2** [comment]
> @BenWilson2
> 
> 1. I will add the patch.
> 2. if you suggestion is to use `from mlflow import MlflowClient` in the entire code base, then it results in import issues like below traceback. Let me check this issue again and circle back.
> 
> ```
> Traceback (most recent call last):
>   File "/opt/hostedtoolcache/Python/3.7.12/x64/bin/mlflow", line 5, in <module>
>     from mlflow.cli import cli
>   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/mlflow/__init__.py", line 32, in <module>
>     import mlflow.tracking._model_registry.fluent
>   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/mlflow/tracking/__init__.py", line 20, in <module>
>     from mlflow.tracking.fluent import _EXPERIMENT_ID_ENV_VAR, _EXPERIMENT_NAME_ENV_VAR, _RUN_ID_ENV_VAR
>   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/mlflow/tracking/fluent.py", line 25, in <module>
>     from mlflow.tracking.default_experiment import registry as default_experiment_registry
>   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/mlflow/tracking/default_experiment/registry.py", line 5, in <module>
>     from mlflow.tracking.default_experiment.databricks_notebook_experiment_provider import (
>   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/mlflow/tracking/default_experiment/databricks_notebook_experiment_provider.py", line 3, in <module>
>     from mlflow import MlflowClient
> ImportError: cannot import name 'MlflowClient' from 'mlflow' (/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/mlflow/__init__.py)
> ```

Yep that'll be the exception :) no circular imports within the init that defines the alias. We'll need to do the direct import there to the lightweight wrapper in `mlflow.tracking.MlflowClient` that points to the implementation in `mlflow.tracking._tracking_service.client.TrackingServiceClient`

**Comment by subramaniam02** [comment]
autoformat

File: mlflow/pyspark/ml/__init__.py [hunk 2]
```diff
@@ -6,8 +6,8 @@
 import weakref
 
 import mlflow
-from mlflow.entities import Metric, Param
 from mlflow.tracking.client import MlflowClient
```
Comment (BenWilson2) [inline mlflow/pyspark/ml/__init__.py:9]: This one was giving some initialization issues?

File: mlflow/store/artifact/models_artifact_repo.py [hunk 3]
```diff
@@ -59,12 +59,11 @@ def split_models_uri(uri):
     def get_underlying_uri(uri):
         # Note: to support a registry URI that is different from the tracking URI here,
         # we'll need to add setting of registry URIs via environment variables.
-        from mlflow.tracking import MlflowClient
```
Comment (BenWilson2) [inline mlflow/store/artifact/models_artifact_repo.py:62]: For local imports, let's not try to import from the module level. We can change the reference here to `from mlflow import MlflowClient` but the local import is desired here.

File: mlflow/store/artifact/models_artifact_repo.py [hunk 4]
```diff
@@ -59,12 +59,11 @@ def split_models_uri(uri):
...
 
         databricks_profile_uri = (
             get_databricks_profile_uri_from_artifact_uri(uri) or mlflow.get_registry_uri()
         )
-        client = MlflowClient(registry_uri=databricks_profile_uri)
+        client = mlflow.MlflowClient(registry_uri=databricks_profile_uri)
```
Comment (BenWilson2) [inline mlflow/store/artifact/models_artifact_repo.py:66]: change to refer to the local import statement

File: mlflow/utils/autologging_utils/client.py [hunk 5]
```diff
@@ -29,6 +28,7 @@
     MAX_METRICS_PER_BATCH,
 )
 
+from mlflow.tracking.client import MlflowClient
```
Comment (BenWilson2) [inline mlflow/utils/autologging_utils/client.py:31]: The locations where the import statement is simply moved, can we revert those?

File: mlflow/pyspark/ml/__init__.py (hunk 2 above)
Comment (subramaniam02) [reply mlflow/pyspark/ml/__init__.py:9]: This is a mistake on my side, will revert it. not sure how tests cases were passing.

**Comment by subramaniam02** [comment]
> This one was giving some initialization issues? 

are you asking if using `from mlflow.tracking.client import MlflowClient` and not `from mlflow import MlflowClient` is due to initialization error/issue ?

**Review by BenWilson2 — APPROVED** [review]
LGTM and thank you!

**Comment by BenWilson2** [comment]
@dbczumar for a second set of eyes on so many changes

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @subramaniam02 !

**Pull request merged**